test/puppetlabs/trapperkeeper/testutils/logging.clj

```diff
@@ -1,28 +1,321 @@
+  original appenders.  If logger-id is not a class, it will be
+  converted via str."
+  [logger-id appenders & body]
+  `(call-with-log-appenders ~logger-id ~appenders (fn [] ~@body)))
+
+(defn call-with-log-event-listener
+  "Calls (listen event) for each logger-id event produced during the
+  evaluation of f.  If logger-id is not a class, it will be converted
+  via str."
+  [logger-id listen f]
+  (with-started [listener (doto (log-event-listener listen) .start)]
+    (call-with-additional-log-appenders logger-id [listener] f)))
+
+(defmacro with-log-event-listener
+  "Calls (listen event) for each logger-id event produced during the
+  evaluation of body.  If logger-id is not a class, it will be
+  converted via str."
+  [logger-id listen & body]
+  `(call-with-log-event-listener ~logger-id ~listen (fn [] ~@body)))
+
+(defmacro with-logging-to-atom
+  "Conjoins all logger-id events produced during the evaluation of the
+  body onto the collection in the destination atom.  If logger-id is
+  not a class, it will be converted via str."
+  [logger-id destination & body]
+  `(with-log-event-listener
+     ~logger-id
+     (fn [event#] (swap! ~destination conj event#))
+     ~@body))
+
+(defn- suppressing-file-appender
+  [log-path]
+  (let [pattern "%-4relative [%thread] %-5level %logger{35} - %msg%n"
+        context (LoggerFactory/getILoggerFactory)]
+    (doto (FileAppender.)
+      (.setFile log-path)
+      (.setAppend true)
+      (.setEncoder (doto (PatternLayoutEncoder.)
+                     (.setPattern pattern)
+                     (.setContext context)
+                     (.start)))
+      (.setContext context)
+      (.start))))
+
+(defn- call-with-log-suppressed-unless-notable [notable-event? f]
+  (let [problem (atom false)
+        log-path (fs/absolute-path (fs/temp-file "tk-suppressed" ".log"))]
+    (try
+      (with-started [appender (suppressing-file-appender log-path)
+                     detector (doto (log-event-listener
+                                     (fn [event]
+                                       (when (notable-event? event)
+                                         (reset! problem true))))
+                                .start)]
+        (with-log-appenders root-logger-name
+          [appender detector]
+          (f)))
+      (finally
+        (if @problem
+          (binding [*out* *err*]
+            (print (slurp log-path))
+            (println "From error log:" log-path))
+          (fs/delete log-path))))))
+
+(defmacro with-log-suppressed-unless-notable
+  "Executes the body with all logging suppressed.  If a notable error
+  is logged, dumps the full log to *err* along with its path.  Assumes
+  that the logging level is already set as desired.  This may not work
+  correctly if the system logback config is altered during the
+  execution of the body."
+  [notable-event? & body]
+  `(#'puppetlabs.trapperkeeper.testutils.logging/call-with-log-suppressed-unless-notable
```

Comment:
Though actually, I should change this, since I think that with-foo macros should often (normally) provide a parallel, supporting call-with-foo function.  I'll make the call-with- public.